src/codec/length_delimited.rs

```diff
@@ -578,6 +578,12 @@ impl Encoder for LengthDelimitedCodec {
             "provided length would overflow after adjustment",
         ))?;
 
+        // If there's no room to write the length field in `dst`, reserve more
+        // capacity so we don't overflow the buffer.
+        if dst.remaining_mut() < self.builder.length_field_len {
+            dst.reserve(self.builder.length_field_len);
```

Comment:
As a potential optimization, we might want to reserve enough buffer capacity for both the length field _and_ the frame here. 

The call to `BytesMut::extend_from_slice` on [line 589](https://github.com/tokio-rs/tokio/pull/682/commits/63a31bc59c86d7b7972cc8b61af8e5af37d88d89#diff-1a8c3780aad5ae7838c4785b000e9d8cR594) later on won't panic, as that method will resize the buffer if there's insufficient capacity to fit the frame. However, if we grow the buffer enough to fit the frame here, we might be able to do a single allocation rather than two?

Reply:
I think doing the one allocation is preferential to (potentially) two.

Reply:
I agree, since the total needed size is known, it should be reserved up front.

Also, you do not need to guard the `reserve` function.

Reply:
I agree, since the total needed size is known, it should be reserved up front.

Also, you do not need to guard the `reserve` function.

Reply:
Okay, thanks @carllerche and @tobz for confirming my suspicion that reserving all the necessary capacity here was the Right Thing :smile: --- I've made that change in 

Also, I think that since we're reserving sufficient capacity here, the `BytesMut::extend_from_slice` on [line 583]( https://github.com/tokio-rs/tokio/pull/682/files/47a32d974063fc8cc55c86a8c82104fe4cfd5d66..63a31bc59c86d7b7972cc8b61af8e5af37d88d89#diff-1a8c3780aad5ae7838c4785b000e9d8cR594) _could_ just be changed to a `BufMut::push_slice` --- @carllerche, is there significant additional overhead in calling `extend_from_slice` when the `BytesMut` has sufficient capacity for the slice?